### ISSUE TITLE
`offset_of` returns a temporary

### ISSUE THREAD

**Issue by tgross35** [rust-lang/rust#124478]
I tried this code:

```rust
struct Foo {
    a: u32,
}

impl Foo {
    fn bar() -> &'static [usize] {
        &[std::mem::offset_of!(Self, a)]
    }
}
```

[Playground link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=91f57feb9b8919f9fb5f115a272ea6da)

I expected to see this happen: no error

Instead, this happened:

```text
error[E0515]: cannot return reference to temporary value
 --> src/lib.rs:7:9
  |
7 |         &[std::mem::offset_of!(Self, a)]
  |         ^-------------------------------
  |         ||
  |         |temporary value created here
  |         returns a reference to data owned by the current function

For more information about this error, try `rustc --explain E0515`.
error: could not compile `playground` (lib) due to 1 previous error
```

### Meta

rustc 1.79 nightly 2024-04-27

**Comment by tgross35** [comment]
I can't think of any reason why it would create the temporary here, considering you can assign the result to a `const` just fine.

cc @GKFX who worked on this most recently

@rustbot label +T-compiler +F-offset_of -needs-triage

**Comment by GKFX** [comment]
It's the `core::hint::must_use` that does it: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=bd518763786b97f87434d5c986cc52cf. `builtin # offset_of` alone doesn't cause the problem.

I'm not sure what the original reason was for making this `must_use`, or whether it would be OK to just remove it? Wrapping the macro in `const {}` works, but that would interfere with array indexing in offset_of if that ever gets supported.
```rust
let x = offset_of!(S, a[y]); // not const.
```

**Comment by asquared31415** [comment]
making it `must_use` makes sense, but if it's not possible to keep that behavior and make it not a temporary, i think it's better UX to remove the `must_use` hint.
